### PR TITLE
[SPARK-22642][SQL] the createdTempDir will not be deleted if an exception occurs, should delete it with try-finally.

### PR DESCRIPTION
**Pull request by zuotingbing**
## What changes were proposed in this pull request?

We found staging directories will not be dropped sometimes in our production environment.
The createdTempDir will not be deleted if an exception occurs, we should delete createdTempDir with try-finally.

This PR is follow-up SPARK-18703.

## How was this patch tested?

exist tests

### CODE REVIEW

**Comment by srowen** [comment]
It's a very long method, and a try-finally around it is hard to follow. Can this wrapper live in the caller? is there another place to ensure something's cleaned up?

**Comment by guoxiaolongzte** [comment]
+1, avoid unpredictable exceptions that cause the temporary directory or file to be deleted.

**Comment by HyukjinKwon** [comment]
Could we minimise the diff? It seems hard to follow to me too.

**Comment by zuotingbing** [comment]
Yes i am also confused why the diff is so big but it  reported by git originally.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 1]
```diff
@@ -104,147 +105,153 @@ case class InsertIntoHiveTable(
...
-          val inheritTableSpecs = true
-          externalCatalog.loadPartition(
-            table.database,
-            table.identifier.table,
-            tmpLocation.toString,
-            partitionSpec,
-            isOverwrite = doHiveOverwrite,
-            inheritTableSpecs = inheritTableSpecs,
-            isSrcLocal = false)
+            // inheritTableSpecs is set to true. It should be set to false for an IMPORT query
```
Comment (wangyum) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: l
          l
The number of spaces is different

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 2]
```diff
@@ -104,147 +105,153 @@ case class InsertIntoHiveTable(
...
-           |Requested partitions: ${partition.keys.mkString(",")}
-           |Table partitions: ${table.partitionColumnNames.mkString(",")}""".stripMargin)
-    }
-
-    // Validate partition spec if there exist any dynamic partitions
-    if (numDynamicPartitions > 0) {
-      // Report error if dynamic partitioning is not enabled
-      if (!hadoopConf.get("hive.exec.dynamic.partition", "true").toBoolean) {
-        throw new SparkException(ErrorMsg.DYNAMIC_PARTITION_DISABLED.getMsg)
+    def processInsert = {
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Generally, we do not like nested method.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 3]
```diff
@@ -104,147 +105,153 @@ case class InsertIntoHiveTable(
...
-        tmpLocation.toString, // TODO: URI
-        overwrite,
-        isSrcLocal = false)
     }
 
-    // Attempt to delete the staging directory and the inclusive files. If failed, the files are
-    // expected to be dropped at the normal termination of VM since deleteOnExit is used.
-    deleteExternalTmpPath(hadoopConf)
+    try {
+      processInsert
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Please create a separate function from line 165 to line 235.

**Comment by zuotingbing** [comment]
I extract a separate function but it has too many parameters.  Could i extract several separate functions?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 4]
```diff
@@ -104,14 +105,61 @@ case class InsertIntoHiveTable(
...
+
+      validateBucketSpec(hadoopConf)
+
+      val partitionAttributes = partitionColumnNames.takeRight(numDynamicPartitions).map { name =>
+        query.resolve(name :: Nil, sparkSession.sessionState.analyzer.resolver).getOrElse {
+          throw new AnalysisException(
+            s"Unable to resolve $name given [${query.output.map(_.name).mkString(", ")}]")
+        }.asInstanceOf[Attribute]
+      }
+
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]:
```Scala
  try {
```

should start from this line, right?

No need to create `validatePartitionSpec ` and `validateBucketSpec ` in this PR. We want to minimized the code changes.

**Comment by zuotingbing** [comment]
please review it again, thanks all.

**Comment by gatorsmile** [comment]
ok to test

**Comment by zuotingbing** [comment]
It would be great if someone help to merge it to master, Thanks! @gatorsmile @srowen @HyukjinKwon @wangyum

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by zuotingbing** [comment]
Thanks srowen.
